src/codegen/llvm/codegen_llvm_visitor.cpp

```diff
@@ -152,6 +152,88 @@ void CodegenLLVMVisitor::emit_procedure_or_function_declaration(const ast::Block
                            *module);
 }
 
+llvm::Value* CodegenLLVMVisitor::visit_arithmetic_bin_op(llvm::Value* lhs,
+                                                         llvm::Value* rhs,
+                                                         unsigned op) {
+    const auto& bin_op = static_cast<ast::BinaryOp>(op);
+    llvm::Value* result;
+
+    switch (bin_op) {
+#define DISPATCH(binary_op, llvm_fp_op, llvm_int_op) \
+    case binary_op:                                  \
+        if (lhs->getType()->isDoubleTy())            \
```

Comment:
Should here be `lhs->getType()->isDoubleTy() || rhs->getType()->isDoubleTy()` ? 🤔

Reply:
Not really, I think. LLVM instructions, particularly binary ones, operate on the same type operands. Therefore to find out what instruction should we use (integer or floating-point) can be determined by a single lhs or rhs. 

In NMODL, I assume that double and integer values never really interact with each other? Doubles are used in computations, as arguments and return values, whereas integers are used for loops/conditions? If they do interact, then we should also emit instructions like `sitofp` to cast types. But I think you have previously mentioned that should not be the case?

Alternatively, I can add a check for both types to be the same. Relying on a single lhs is not that safe, I agree.

Reply:
>  assume that double and integer values never really interact with each other? Doubles are used in computations, as arguments and return values, whereas integers are used for loops/conditions?

thats correct

> But I think you have previously mentioned that should not be the case?

yes. I was thinking that _at some point_ we might emit `1+3` as integer operation but then I realised that people do write `1/3` in DSL and there we don't want to bring integer types.

In summary, ignore my previous comment and emit floating point instruction.

One suggestion though: for floating point type, can you put some flag in the class indicating floating point type is `double` or `float`? e.g. `bool floating_type_double = true or false;`. If we toggle this, code generation should emit either `FloatType` or `DoubleType` ? (assuming its not too much work). This would be helpful for evaluations / benchmarking later.

Reply:
> One suggestion though: for floating point type, can you put some flag in the class indicating floating point type is double or float? e.g. bool floating_type_double = true or false;. If we toggle this, code generation should emit either FloatType or DoubleType ? (assuming its not too much work). This would be helpful for evaluations / benchmarking later.

Sure, will add this!